### PR TITLE
add blog post: 50 ways to crash our product

### PR DESCRIPTION
**Pull request by llogiq**
"I give up. I've found 1 way to crash our product with an unintended error, reproduced some known limitations and outlined a number of error messages we can improve on. I fell far short of my original goal, which either means I'm really bad at finding errors, or our code is incredibly reliable. Given the track record of software written in Rust, I'd like to think it's the latter, but I'll leave judgement to you."

### CODE REVIEW

File: docs/blog/2021-09-27-crash.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+image: https://commons.wikimedia.org/wiki/File:Dialog-error-round.svg#/media/File:Dialog-error-round.svg
+hide_table_of_contents: false
+---
+
+I personally think that the software we build should make more people's lives better than it makes worse. So when users recently started filing bug reports, I read them with mixed feelings. On one hand, it meant that those particular users were actually using [synth](https://getsynth.com), on the other hand, it also meant that we were failing to give them a polished experience. So when it was my turn to write more stuff about what we do here, I set myself a challenge: Find as many ways as I can to break our product.
+
+I briefly considered fuzzing, but decided against it. It felt like cheating. Where's the challenge in that? Also, I wanted to be sure that the bugs would be reachable by ordinary (or perhaps at least exceptional) users, and would accept misleading error messages (that a fuzzer couldn't well decidde) as bugs. Finally I am convinced I learn more about some code when actively trying to break it, and that's always a plus. So "let's get cracking!" I quoth and off I went.
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:15]: decidde -> decide

File: docs/blog/2021-09-27-crash.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+### Overview
+
+Before we start, I should perhaps consider giving a short architectural overview on synth. Basically the software has four parts:
+
+1. The DSL (which is implemented by a set of types in `core/src/schema` that get deserialized from JSON),
+2. a compiler that creates a `graph` (which form a directed acyclic graph of items that can generate values),
+3. export and
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:23]: export (writing to the data sink)

File: docs/blog/2021-09-27-crash.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+4. import facilitis (for database connectivity)
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:24]: facilities* (reading from the database)

File: docs/blog/2021-09-27-crash.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+title: 50 ways to crash our product
+author: Andre Bogus 
+author_title: Chief Rustacean
+author_url: https://github.com/getsynth
+author_image_url: https://avatars.githubusercontent.com/u/4200835?v=4
+tags: [synth, testing, story]
+description: Trying to break our own code for fun and profit
+image: https://commons.wikimedia.org/wiki/File:Dialog-error-round.svg#/media/File:Dialog-error-round.svg
+hide_table_of_contents: false
+---
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:11]: I would add an image at the top to a: make the content more engaging, and b: better thumbnail for the link. Maybe something like this -> 
image
?

File: docs/blog/2021-09-27-crash.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+CREATE TABLE IF NOT EXISTS json (
+    data JSON
+);
+
+DELETE FROM json;
+
+INSERT INTO json (data) VALUES ('{ "a": ["b", 42] }');
+```
+
+Now call `synth import fail --from mysql://<user>:<password>@<host>:<port>/<database>` to get
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:269]: The `fail` namespace name may be misleading here?

File: docs/blog/2021-09-27-crash.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+will generate the following error:
+
+```console
+Error: Unable to open the namespace
+
+Caused by:
+    0: at file 12_stack_depth/stack_depth.json
+    1: Failed to parse collection
+    2: recursion limit exceeded at line 1 column 2929
+```
```
Comment (christoshadjiaslanis) [inline docs/blog/2021-09-27-crash.md:324]: hah! that's cool :D

**Review by christoshadjiaslanis — COMMENTED** [review]
Really nice @llogiq. Well written and fun to read (for me at least :shrug: ). Left a few minor comments for you.

File: docs/blog/2021-09-27-crash.md (hunk 5 above)
Comment (llogiq) [reply docs/blog/2021-09-27-crash.md:269]: Ok, I'll use `jsonnamespace` instead.

File: docs/blog/2021-09-27-crash.md (hunk 6 above)
Comment (llogiq) [reply docs/blog/2021-09-27-crash.md:324]: I knew I had to overflow the stack at least once.

**Pull request merged**